### PR TITLE
Cornerstone client

### PR DESCRIPTION
**Pull request by devsprite**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,8 @@
 
 \.idea/
+
+node_modules/
+
+test/tests\.test\.js
```
Comment (bouiboui) [inline .gitignore:6]: You should add a `*` to catch future test files

File: src/cornerstone.js [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+class Cornerstone {
+
+  constructor({apiId, apiSecret, username, alias, corpname}) {
+    this.apiId     = apiId;
+    this.apiSecret = apiSecret;
+    this.username  = username;
+    this.alias     = alias;
+    this.corpname  = corpname;
+    this.dateTime  = this.getDatetimeUTC();
```
Comment (bouiboui) [inline src/cornerstone.js:13]: Should this really happen in the constructor or when you initialize the connection?

File: src/cornerstone.js [hunk 3]
```diff
@@ -0,0 +1,241 @@
+const config      = require('../config/common')();
+const Auth        = require('./lib/Auth');
+const querystring = require('querystring');
+
+class Cornerstone {
+
+  constructor({apiId, apiSecret, username, alias, corpname}) {
```
Comment (bouiboui) [inline src/cornerstone.js:7]: Please add JSDoc to every method!

File: src/cornerstone.js [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    this.apiId     = apiId;
+    this.apiSecret = apiSecret;
+    this.username  = username;
+    this.alias     = alias;
+    this.corpname  = corpname;
+    this.dateTime  = this.getDatetimeUTC();
+    this.auth      = null;
+  }
+
+  setAuth() {
```
Comment (bouiboui) [inline src/cornerstone.js:21]: This method should be named `getAuth`, and get used as such.

File: src/lib/Auth.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  /**
+   * Get base url
+   * @param corpname {string} fice, fice-pilot, fice-stage
+   * @returns {string} url
+   */
+  getBaseUrl({corpname}) {
+    return config.CORNERSTONE_BASE_URL.replace('{corpname}', corpname);
+  }
```
Comment (bouiboui) [inline src/lib/Auth.js:171]: You're coupling this code to the config, which is a bad thing.
You should ask for the baseUrl in the constructor instead, and this code (the `.replace`) should happen one level higher.

File: src/lib/Auth.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  getSignatureSession({method, sessionToken, sessionSecret, httpUrl, dateUTC}) {
+    const sessionTokenKey = 'x-csod-session-token:' + sessionToken;
+    const dateCsod        = 'x-csod-date:' + dateUTC;
+    const stringToSign    = method + '\n' + dateCsod + '\n' + sessionTokenKey + '\n' + httpUrl;
+    const secretKey       = new Buffer(sessionSecret, 'base64');
+    const hmac            = crypto.createHmac('sha512', secretKey);
+    console.log('[getSignatureSession] - signature: ', JSON.stringify(stringToSign));
+
+    return hmac.update(stringToSign).digest('base64');
+  }
```
Comment (bouiboui) [inline src/lib/Auth.js:55]: `getSignature` and `getSignatureSession` share a lot of code that could be abstracted away in a third method.

File: src/cornerstone.js [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    try {
+      const userObject = await connectionSession.get(path);
+      if (userObject.status === 200 || userObject.status === 201) {
+        console.log('[getReporting] - response: ', userObject.data);
+        return userObject.data.value[0];
+      }
+    } catch (e) {
+      console.log('[getReporting] - Error:', e.response.data);
+    }
+    console.log('erere');
```
Comment (bouiboui) [inline src/cornerstone.js]: Please remove this

File: src/cornerstone.js [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+    return await this.getReporting({request: request, url: url});
+  }
+
+  async getReportingKeycodeByUserRef({user_ref}) {
+    const request = "?$filter=tu_training_unit_key_code eq '{user_ref}'".replace('{user_ref}', user_ref);
+    const url     = config.CORNERSTONE_REPORTING_KEYCODE;
+
+    return await this.getReporting({request: request, url: url});
+  }
```
Comment (bouiboui) [inline src/cornerstone.js:163]: There's a lot of duplicated code that could be moved to `getReporting`.
Ex:
```js
async getReportingKeycodeByUserRef({user_ref}) {
  return await this.getReporting({
    url: config.CORNERSTONE_REPORTING_KEYCODE
    filters: [{ tu_training_unit_key_code: { eq: user_ref } }], 
  });
}
```

File: src/lib/Auth.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    if (sessionFile) {
+      const dateNow     = new Date();
+      const dateSession = new Date(sessionFile.expiresOn);
+
+      if (dateNow < dateSession) {
+        return sessionFile;
+      }
+    }
+
+    const httpUrl  = config.CORNERSTONE_PATH_SESSION;
```
Comment (bouiboui) [inline src/lib/Auth.js:115]: This should come from the constructor.

File: src/lib/Auth.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+      apiSecret: this.apiSecret,
+      httpUrl:   httpUrl,
+      dateUTC:   dateUTC
+    });
+
+    const baseUrl = this.getBaseUrl({corpname: this.corpname});
+
+    const path = baseUrl + httpUrl + '?userName={username}&alias={alias}'
+      .replace('{username}', this.username)
+      .replace('{alias}', this.alias + Date.now());
```
Comment (bouiboui) [inline src/lib/Auth.js]: You shoud use template strings for this!
```js
const path = `${baseUrl}${httpUrl}?userName=${username}&alias=${this.alias + Date.now()}';
```

File: src/lib/Auth.js [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+  readSession() {
+    let file;
+    if (fs.existsSync(config.TMP_PATH + 'session.json')) {
+      file = fs.readFileSync(config.TMP_PATH + 'session.json', 'utf8');
+      console.log('[readSession] - session tmp file: ', file);
+
+      return file;
+    }
+    file && console.log('3', file);
```
Comment (bouiboui) [inline src/lib/Auth.js]: '3'?

File: .gitignore (hunk 1 above)
Comment (devsprite) [reply .gitignore:6]: I want catch only test.js because there are credentials in this file.

File: src/cornerstone.js (hunk 2 above)
Comment (devsprite) [reply src/cornerstone.js:13]: Only in constructor because I want dateTime unique

File: src/cornerstone.js (hunk 4 above)
Comment (devsprite) [reply src/cornerstone.js:21]: are you sure ?

File: src/cornerstone.js (hunk 7 above)
Comment (devsprite) [reply src/cornerstone.js]: it's done

File: src/cornerstone.js (hunk 8 above)
Comment (devsprite) [reply src/cornerstone.js:163]: the requests  are not only filter, there are more possibility.

File: src/lib/Auth.js (hunk 10 above)
Comment (devsprite) [reply src/lib/Auth.js]: You're right, it's more eleguant. It's done

File: src/lib/Auth.js (hunk 11 above)
Comment (devsprite) [reply src/lib/Auth.js]: it's done

File: src/cornerstone.js (hunk 4 above)
Comment (bouiboui) [reply src/cornerstone.js:21]: No 😋 Are there going to be different instances with different values? If not, it should be a singleton (instantiated once)

File: src/cornerstone.js (hunk 8 above)
Comment (bouiboui) [reply src/cornerstone.js:163]: Show me where :)

**Pull request merged**